bigframes/functions/remote_function_template.py

```diff
@@ -0,0 +1,222 @@
+# The number and the order of values in each row is expected to exactly
+# match to the number and order of arguments in the udf , e.g. if the udf is
+#   def foo(x: int, y: str):
+#     ...
+# then the http request body could look like
+# {
+#   ...
+#   "calls" : [
+#     [123, "hello"],
+#     [456, "world"]
+#   ]
+#   ...
+# }
+# https://cloud.google.com/bigquery/docs/reference/standard-sql/remote-functions#input_format
+def udf_http(request):
+    import json
+
+    from flask import jsonify
+
+    try:
+        request_json = request.get_json(silent=True)
+        calls = request_json["calls"]
+        replies = []
+        for call in calls:
+            reply = udf(*call)
+            replies.append(reply)
+        return_json = json.dumps({"replies": replies})
+        return return_json
+    except Exception as e:
+        return jsonify({"errorMessage": str(e)}), 400
+
+
+def udf_http_row_processor(request):
+    import json
+    import math
+
+    from flask import jsonify
+    import pandas as pd
+
+    try:
+        request_json = request.get_json(silent=True)
+        calls = request_json["calls"]
+        replies = []
+        for call in calls:
+            reply = udf(get_pd_series(call[0]))
+            if isinstance(reply, float) and (math.isnan(reply) or math.isinf(reply)):
+                # json serialization of the special float values (nan, inf, -inf)
+                # is not in strict compliance of the JSON specification
+                # https://docs.python.org/3/library/json.html#basic-usage.
+                # Let's convert them to a quoted string representation ("NaN",
+                # "Infinity", "-Infinity" respectively) which is handled by
+                # BigQuery
+                reply = json.dumps(reply)
+            elif pd.isna(reply):
+                # Pandas N/A values are not json serializable, so use a python
+                # equivalent instead
+                reply = None
+            elif hasattr(reply, "item"):
+                # Numpy types are not json serializable, so use its Python
+                # value instead
+                reply = reply.item()
+            replies.append(reply)
+        return_json = json.dumps({"replies": replies})
+        return return_json
+    except Exception as e:
+        return jsonify({"errorMessage": str(e)}), 400
+
+
+def generate_udf_code(def_, directory):
+    """Generate serialized bytecode using cloudpickle given a udf."""
+    udf_code_file_name = "udf.py"
+    udf_bytecode_file_name = "udf.cloudpickle"
+
+    # original code, only for debugging purpose
+    udf_code = textwrap.dedent(inspect.getsource(def_))
+    udf_code_file_path = os.path.join(directory, udf_code_file_name)
+    with open(udf_code_file_path, "w") as f:
+        f.write(udf_code)
+
+    # serialized bytecode
+    udf_bytecode_file_path = os.path.join(directory, udf_bytecode_file_name)
+    with open(udf_bytecode_file_path, "wb") as f:
```

Comment:
Do we really need files on disk or File Objects in memory are enough? Potentially better performance.

Reply:
I can try this out in a future PR. I'd like to keep this a pure refactor without any behavior changes. Filed issue 345433300.

```suggestion
    # TODO(b/345433300): try io.BytesIO to avoid writing to the file system
    with open(udf_bytecode_file_path, "wb") as f:
```